### PR TITLE
Move unneeded_field_pattern to restriction group

### LINKED ISSUE #1741: unneeded_field_pattern
```
warning: You matched a field with a wildcard pattern. Consider using `..` instead, #[warn(unneeded_field_pattern)] on by default
   --> src/engine/phases/incremental_sync_step.rs:398:55
    |
398 |         let UpdateHiddenStateReq{root_relative_paths, infinitize_unignores: _} = req;
    |                                                       ^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: Try with `UpdateHiddenStateReq { root_relative_paths, .. }`
    = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#unneeded_field_pattern
```

Here is an an exact warning from our code. We have a large codebase (several tens of thousands of lines) and get a lot of refactor value out of having a complete destructure without .. syntax.

When we use {..} it increases the probability that someone changes the fields of UpdateHiddenStateReq and forgets to change the usage site here.

One of rust's biggest advantages for us is how it prevents us from shooting ourselves in the foot during refactors or edits in situations like this.

Now I understand that not everyone has this concern and the .. syntax definitely has its benefits, but I do think it's worth discussing if this should be a warning. What do you think?

Thanks

### PR DESCRIPTION
**Pull request by nipunn1313**
Fixes #1741

changelog: Move unneeded_field_pattern to pedantic group

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ document.
 
 [69f99e7...master](https://github.com/rust-lang/rust-clippy/compare/69f99e7...master)
 
+* Move `unneeded_field_pattern` to pedantic group [#1741](https://github.com/rust-lang/rust-clippy/issues/1741)
+
```
Comment (JohnTitor) [inline CHANGELOG.md]: You don't need to update CHANGELOG in this PR, it's usually updated in release time.

Comment (nipunn1313) [reply CHANGELOG.md]: sgtm! Updated w/o this change.

File: clippy_lints/src/misc_early.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ declare_clippy_lint! {
     /// let { a: _, b: ref b, c: _ } = ..
     /// ```
     pub UNNEEDED_FIELD_PATTERN,
-    style,
+    pedantic,
```
Comment (flip1995) [inline clippy_lints/src/misc_early.rs]: I agree with @gnzlbg , that this should probably even be a `restriction` lint, so that you have to opt-in.

Comment (nipunn1313) [reply clippy_lints/src/misc_early.rs]: I can make that change if you request it.

Pedantic is described as
```
`clippy::pedantic` (lints which are rather strict, off by default)
```

Restriction is described as
```
The [lint list](https://rust-lang.github.io/rust-clippy/master/index.html) also contains "restriction lints", which are
for things which are usually not considered "bad", but may be useful to turn on in specific cases. These should be used
very selectively, if at all.
```

I'll switch it to restriction which seems to match this better.

Comment (flip1995) [reply clippy_lints/src/misc_early.rs]: > which are usually not considered "bad"

I think this perfectly matches this lint 👍

**Pull request merged**